File: app/src/main/java/com/minirogue/starwarscanontracker/view/fragment/MediaListFragment.java
package com.minirogue.starwarscanontracker.view.fragment;

import android.content.Context;
import android.os.Bundle;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.PopupMenu;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.minirogue.starwarscanontracker.R;
import com.minirogue.starwarscanontracker.application.CanonTrackerApplication;
import com.minirogue.starwarscanontracker.model.SortStyle;
import com.minirogue.starwarscanontracker.model.room.entity.MediaNotes;
import com.minirogue.starwarscanontracker.model.room.pojo.FullFilter;
import com.minirogue.starwarscanontracker.view.FilterChip;
import com.minirogue.starwarscanontracker.view.adapter.SWMListAdapter;
import com.minirogue.starwarscanontracker.viewmodel.MediaListViewModel;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import javax.inject.Inject;

import me.zhanghai.android.fastscroll.FastScrollerBuilder;

public class MediaListFragment extends Fragment {
    private static final String TAG = "Media List";

    private SWMListAdapter adapter;
    private MediaListViewModel mediaListViewModel;
    private ChipGroup chipGroup;
    private Chip sortChip;
    private FloatingActionButton sortFAB;
    private Context ctx;

    private SWMListAdapter.AdapterInterface adapterInterface = new SWMListAdapter.AdapterInterface() {
        @Override
        public void onItemClicked(int itemId) {
            ViewMediaItemFragment viewMediaItemFragment = new ViewMediaItemFragment();
            Bundle bundle = new Bundle();
            bundle.putInt(getString(R.string.bundleItemId), itemId);
            viewMediaItemFragment.setArguments(bundle);
            getActivity().getSupportFragmentManager().beginTransaction()
                    .replace(R.id.fragment_container, viewMediaItemFragment)
                    .addToBackStack(null)
                    .commit();//TODO this should be handled by the activity
        }

        @Override
        public void onCheckbox1Clicked(MediaNotes mediaNotes) {
            mediaNotes.flipCheck1();
            mediaListViewModel.update(mediaNotes);
        }

        @Override
        public void onCheckbox2Clicked(MediaNotes mediaNotes) {
            mediaNotes.flipCheck2();
            mediaListViewModel.update(mediaNotes);
        }

        @Override
        public void onCheckbox3Clicked(MediaNotes mediaNotes) {
            mediaNotes.flipCheck3();
            mediaListViewModel.update(mediaNotes);
        }

        @NotNull
        @Override
        public String getMediaTypeString(int mediaTypeId) {
            return mediaListViewModel.convertTypeToString(mediaTypeId);
        }

        @NotNull
        @Override
        public String getSeriesString(int seriesId) {
            //TODO
            return "Series not found";
        }

        @Override
        public boolean isNetworkAllowed() {
            return mediaListViewModel.isNetworkAllowed();
        }
    };

    @Inject
    ViewModelProvider.Factory viewModelFactory;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View fragmentView = inflater.inflate(R.layout.fragment_media_list, container, false);
        ctx = fragmentView.getContext();
        ((CanonTrackerApplication) getActivity().getApplication()).appComponent.inject(this);
        mediaListViewModel = ViewModelProviders.of(this, viewModelFactory).get(MediaListViewModel.class);
        mediaListViewModel.getFilteredMediaAndNotes().observe(getViewLifecycleOwner(), mediaAndNotes -> adapter.submitList((mediaAndNotes)));

        RecyclerView recyclerView = fragmentView.findViewById(R.id.media_recyclerview);
        chipGroup = fragmentView.findViewById(R.id.filter_chip_group);
        makeCurrentSortChip();
        sortFAB = fragmentView.findViewById(R.id.sort_floating_action_button);
        PopupMenu sortMenu = makeSortMenu();
        sortFAB.setOnClickListener(view -> sortMenu.show());


        RecyclerView.LayoutManager mLayoutManager = new LinearLayoutManager(fragmentView.getContext());
        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(recyclerView.getContext(), LinearLayout.VERTICAL);
        adapter = new SWMListAdapter(adapterInterface);
        recyclerView.setHasFixedSize(true);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(mLayoutManager);
        recyclerView.addItemDecoration(dividerItemDecoration);
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);

            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (dy > 0) {//Scroll down
                    sortFAB.hide();
                } else if (dy < 0) {//Scroll up
                    sortFAB.show();
                }
            }
        });

        new FastScrollerBuilder(recyclerView).build();

        mediaListViewModel.getCheckBoxText().observe(getViewLifecycleOwner(), adapter::updateCheckBoxText);
        mediaListViewModel.getCheckBoxVisibility().observe(getViewLifecycleOwner(), adapter::updateCheckBoxVisible);
        mediaListViewModel.getActiveFilters().observe(getViewLifecycleOwner(), this::setFilterChips);

        return fragmentView;
    }

    private PopupMenu makeSortMenu() {
//        mediaListViewModel.toggleSort();
        PopupMenu newSortMenu;
        newSortMenu = new PopupMenu(ctx, sortFAB, Gravity.END);
        for (int style : SortStyle.Companion.getAllStyles()) {
            newSortMenu.getMenu().add(0, style, 0, SortStyle.Companion.getSortText(style));
        }
        newSortMenu.setOnMenuItemClickListener(menuItem -> {
            mediaListViewModel.setSort(menuItem.getItemId());
            return true;
        });
        return newSortMenu;
    }

    private void setFilterChips(List<FullFilter> filters) {
        chipGroup.removeAllViews();
        chipGroup.addView(sortChip);
        for (FullFilter filter : filters) {
            chipGroup.addView(makeCurrentFilterChip(filter));
        }
    }

    private Chip makeCurrentFilterChip(final FullFilter fullFilter) {
        Chip filterChip = new FilterChip(fullFilter, ctx);
        filterChip.setOnCloseIconClickListener(view -> mediaListViewModel.deactivateFilter(fullFilter.getFilterObject()));
        return filterChip;
    }

    private void makeCurrentSortChip() {
        sortChip = new Chip(ctx);
        sortChip.setChipIconVisible(true);
        sortChip.setOnClickListener(view -> mediaListViewModel.reverseSort());
        mediaListViewModel.getSortStyle().observe(getViewLifecycleOwner(), this::updateSortChip);
        chipGroup.addView(sortChip);
    }

    private void updateSortChip(SortStyle sortStyle) {
        sortChip.setText(sortStyle.getText());
        if (sortStyle.isAscending()) {
            sortChip.setChipIcon(getResources().getDrawable(R.drawable.ic_ascending_sort));
        } else {
            sortChip.setChipIcon(getResources().getDrawable(R.drawable.ic_descending_sort));
        }
    }
}

File: app/src/main/java/com/minirogue/starwarscanontracker/view/fragment/MediaListFragment.kt
package com.minirogue.starwarscanontracker.view.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.chip.Chip
import com.minirogue.starwarscanontracker.R
import com.minirogue.starwarscanontracker.core.model.SortStyle
import com.minirogue.starwarscanontracker.core.model.room.entity.MediaNotesDto
import com.minirogue.starwarscanontracker.databinding.FragmentMediaListBinding
import com.minirogue.starwarscanontracker.view.FilterChip
import com.minirogue.starwarscanontracker.view.adapter.SWMListAdapter
import com.minirogue.starwarscanontracker.view.adapter.SWMListAdapter.AdapterInterface
import com.minirogue.starwarscanontracker.view.viewBinding
import com.minirogue.starwarscanontracker.viewmodel.MediaListViewModel
import dagger.hilt.android.AndroidEntryPoint
import filters.MediaFilter
import kotlinx.coroutines.launch
import me.zhanghai.android.fastscroll.FastScrollerBuilder

@AndroidEntryPoint
class MediaListFragment : Fragment() {
    private val layoutId = R.layout.fragment_media_list
    private val binding by viewBinding(FragmentMediaListBinding::bind)

    private val mediaListViewModel: MediaListViewModel by viewModels()

    private lateinit var sortChip: Chip

    private val adapterInterface: AdapterInterface = object : AdapterInterface {
        override fun onItemClicked(itemId: Int) {
            val viewMediaItemFragment = ViewMediaItemFragment()
            val bundle = Bundle()
            bundle.putInt(getString(R.string.bundleItemId), itemId)
            viewMediaItemFragment.arguments = bundle
            // TODO this should be handled by the activity or swapped to navigation components
            requireActivity().supportFragmentManager.beginTransaction()
                .replace(R.id.fragment_container, viewMediaItemFragment)
                .addToBackStack(null)
                .commit()
        }

        override fun onCheckbox1Clicked(mediaNotesDto: MediaNotesDto) {
            mediaNotesDto.flipCheck1()
            mediaListViewModel.update(mediaNotesDto)
        }

        override fun onCheckbox2Clicked(mediaNotesDto: MediaNotesDto) {
            mediaNotesDto.flipCheck2()
            mediaListViewModel.update(mediaNotesDto)
        }

        override fun onCheckbox3Clicked(mediaNotesDto: MediaNotesDto) {
            mediaNotesDto.flipCheck3()
            mediaListViewModel.update(mediaNotesDto)
        }

        override fun getMediaTypeString(mediaTypeId: Int): String {
            return mediaListViewModel.convertTypeToString(mediaTypeId)
        }

        override fun getSeriesString(seriesId: Int): String {
            // TODO
            return "Series not found"
        }

        override fun isNetworkAllowed(): Boolean {
            return mediaListViewModel.isNetworkAllowed.value
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.media_list_menu, menu)

        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.sort_by_date_menu_item -> {
                mediaListViewModel.setSort(SortStyle.SORT_DATE)
                true
            }

            R.id.sort_by_timeline_menu_item -> {
                mediaListViewModel.setSort(SortStyle.SORT_TIMELINE)
                true
            }

            R.id.sort_by_title_menu_item -> {
                mediaListViewModel.setSort(SortStyle.SORT_TITLE)
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(layoutId, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        sortChip = makeCurrentSortChip()

        with(binding.mediaRecyclerview) {
            layoutManager = LinearLayoutManager(context)
            adapter = SWMListAdapter(adapterInterface).also {
                mediaListViewModel.filteredMediaAndNotes.observe(viewLifecycleOwner) { mediaAndNotes -> it.submitList(mediaAndNotes) }
                viewLifecycleOwner.lifecycleScope.launch {
                    repeatOnLifecycle(Lifecycle.State.STARTED) {
                        launch {
                            mediaListViewModel.checkBoxText.collect { newCheckBoxText -> it.updateCheckBoxText(newCheckBoxText) }
                        }
                        launch {
                            mediaListViewModel.checkBoxVisibility.collect { newIsCheckboxActive ->
                                it.updateCheckBoxVisible(newIsCheckboxActive)
                            }
                        }
                    }
                }
                setHasFixedSize(true)
                addItemDecoration(DividerItemDecoration(context, LinearLayout.VERTICAL))

                FastScrollerBuilder(this).build()
            }
        }

        mediaListViewModel.activeFilters.observe(viewLifecycleOwner) { filters: List<MediaFilter> -> setFilterChips(filters) }
    }

    private fun setFilterChips(filters: List<MediaFilter>) = with(binding.filterChipGroup) {
        removeAllViews()
        addView(sortChip)
        for (filter in filters) {
            addView(makeCurrentFilterChip(filter))
        }
    }

    private fun makeCurrentFilterChip(mediaFilter: MediaFilter): Chip {
        val filterChip: Chip = FilterChip(mediaFilter, requireContext())
        filterChip.setOnCloseIconClickListener { mediaListViewModel.deactivateFilter(mediaFilter) }
        return filterChip
    }

    private fun makeCurrentSortChip(): Chip =
        Chip(context).apply {
            isChipIconVisible = true
            setOnClickListener { mediaListViewModel.reverseSort() }
            mediaListViewModel.sortStyle.observe(
                viewLifecycleOwner,
                { sortStyle: SortStyle -> updateSortChip(sortStyle) }
            )
        }.also {
            binding.filterChipGroup.addView(it)
        }

    private fun updateSortChip(sortStyle: SortStyle) = with(sortChip) {
        text = sortStyle.getText()
        chipIcon = if (sortStyle.isAscending()) {
            ResourcesCompat.getDrawable(resources, R.drawable.ic_ascending_sort, null)
        } else {
            ResourcesCompat.getDrawable(resources, R.drawable.ic_descending_sort, null)
        }
    }
}
